aptos-move/framework/aptos-framework/sources/Voting.move

```diff
@@ -0,0 +1,319 @@
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = Table::borrow_mut(&mut voting_forum.proposals, proposal_id);
+
+        // Check if the proposal can be resolved early.
+        if (Option::is_some(&proposal.early_resolution_vote_threshold)) {
+            let early_resolution_threshold = *Option::borrow(&proposal.early_resolution_vote_threshold);
+            if (proposal.yes_votes >= early_resolution_threshold || proposal.no_votes >= early_resolution_threshold) {
+                return true
+            };
+        };
+
+        // Otherwise, the proposal is only closed if expiration has passed.
+        Timestamp::now_seconds() >= proposal.expiration_secs
+    }
+
+    /// Return the state of the proposal with given id.
+    ///
+    /// @param voting_forum_address The address of the forum where the proposals are stored.
+    /// @param proposal_id The proposal id.
+    /// @return Proposal state as an enum value.
+    public fun get_proposal_state<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+    ): u64 acquires VotingForum {
+        if (is_voting_closed<ProposalType>(voting_forum_address, proposal_id)) {
+            let voting_forum = borrow_global<VotingForum<ProposalType>>(voting_forum_address);
+            let proposal = Table::borrow(&voting_forum.proposals, proposal_id);
+            let yes_votes = proposal.yes_votes;
+            let no_votes = proposal.no_votes;
+
+            if (yes_votes > no_votes && yes_votes + no_votes >= proposal.min_vote_threshold) {
+                PROPOSAL_STATE_SUCCEEDED
+            } else {
+                PROPOSAL_STATE_FAILED
+            }
+        } else {
+            PROPOSAL_STATE_PENDING
+        }
+    }
+
+    public fun get_proposal_expiration_secs<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+    ): u64 acquires VotingForum {
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = Table::borrow_mut(&mut voting_forum.proposals, proposal_id);
+        proposal.expiration_secs
+    }
+
+
+    // TODO: Implement native verify_execution_hash.
+    fun verify_execution_hash(_execution_hash: vector<u8>): bool {
+        true
+    }
+
+    #[test_only]
+    struct TestProposal has store {
+        capability: u64,
+    }
+
+    #[test(core_resources = @CoreResources, governance = @0x123)]
+    public(script) fun test_voting(core_resources: signer, governance: signer) acquires VotingForum {
+        Timestamp::set_time_has_started_for_testing(&core_resources);
+
+        // Register voting forum and create a proposal.
+        register<TestProposal>(&governance);
+        let governance_address = Signer::address_of(&governance);
+        let proposal = TestProposal {
+            capability: 10,
```

Comment:
what is this capability used for ?

Reply:
Good catch - Nothing. It's definitely not needed for the tests. Removed now.